include/dlaf/solver/triangular/impl.h

```diff
@@ -424,6 +470,294 @@ void Triangular<backend, device, T>::call_LLN(comm::CommunicatorGrid grid, blas:
   }
 }
 
+template <Backend backend, Device device, class T>
+void Triangular<backend, device, T>::call_LUN(comm::CommunicatorGrid grid, blas::Diag diag, T alpha,
+                                              Matrix<const T, device>& mat_a, Matrix<T, device>& mat_b) {
+  using namespace triangular_lun;
+
+  auto executor_hp = dlaf::getHpExecutor<backend>();
+  auto executor_np = dlaf::getNpExecutor<backend>();
+  auto executor_mpi = dlaf::getMPIExecutor<backend>();
+
+  // Set up MPI executor pipelines
+  common::Pipeline<comm::Communicator> mpi_row_task_chain(grid.rowCommunicator());
+  common::Pipeline<comm::Communicator> mpi_col_task_chain(grid.colCommunicator());
+
+  const comm::Index2D this_rank = grid.rank();
+
+  const matrix::Distribution& distr_a = mat_a.distribution();
+  const matrix::Distribution& distr_b = mat_b.distribution();
+  auto a_rows = mat_a.nrTiles().rows();
+  auto local_rows = distr_a.localNrTiles().rows();
+  auto b_local_cols = distr_b.localNrTiles().cols();
+
+  // If mat_b is empty return immediately
+  if (mat_b.size().isEmpty())
+    return;
+
+  constexpr std::size_t n_workspaces = 2;
+  common::RoundRobin<matrix::Panel<Coord::Col, T, device>> a_panels(n_workspaces, distr_a);
+  common::RoundRobin<matrix::Panel<Coord::Row, T, device>> b_panels(n_workspaces, distr_b);
+
+  for (SizeType k = a_rows - 1; k > -1; --k) {
+    const GlobalTileIndex kk(k, k);
+    auto kk_rank = distr_a.rankGlobalTile(kk);
+
+    const LocalTileIndex kk_offset{
+        distr_a.nextLocalTileFromGlobalTile<Coord::Row>(k + 1),
+        distr_a.nextLocalTileFromGlobalTile<Coord::Col>(k),
+    };
+
+    const LocalTileIndex bt_offset{distr_b.nextLocalTileFromGlobalTile<Coord::Row>(k), 0};
+
+    auto& a_panel = a_panels.nextResource();
+    auto& b_panel = b_panels.nextResource();
+    if (k == a_rows - 1)
+      a_panel.setWidth(mat_a.tileSize(kk).rows());
+    b_panel.setHeight(mat_a.tileSize(kk).cols());
+
+    if (kk_rank.col() == this_rank.col()) {
+      for (SizeType i_local = kk_offset.row() - 1; i_local > -1; --i_local) {
```

Comment:
In this cases I prefer the condition `i_local >= 0`. @albestro @teonnik @msimberg what do you think?
(Other occurrences in the code)

Reply:
I agree, I think that's clearer.